File: bindings/typescript/src/assemblers/grid.ts
import { FragmentId, GridLayoutFragment } from '@vizstack/schema';
import { FragmentAssembler } from '../fragment-assembler';
import _ from 'lodash';

type GridCell = { row: number; col: number; height: number; width: number };
type GridBounds = { r: number; c: number; R: number; C: number };

export function withinBounds(bounds: GridBounds, row: number, col: number): boolean {
    const { r, c, R, C } = bounds;
    return r <= row && row <= R && c <= col && col <= C;
}

export function parseGridString(str: string): Record<string, GridCell> {
    const rows = str
        .split(/[\n,|]+/)
        .map((row) => row.replace(/\s/g, ''))
        .filter((row) => row.length > 0);
    if (!rows || !rows.every((row) => row.length === rows[0].length)) {
        throw new Error(`Specification string must be rectangular, got rows: ${rows}`);
    }
    const grid: string[][] = rows.map((row) => row.split(''));
    const bounds: Record<string, GridBounds> = {};

    for (let r = 0; r < grid.length; r++) {
        for (let c = 0; c < grid[0].length; c++) {
            const char = grid[r][c];
            if (char in bounds) {
                // A previously encountered character must be within the greedily expanded bounds.
                if (withinBounds(bounds[char], r, c)) continue;
                throw new Error(`Specification string malformed for cell: ${char}`);
            } else {
                if (char === '.') continue;
                // Create a new cell and greedily expand col-wise then row-wise while bounds
                // keep fencing a contiguous rectangle of `char`.
                let R = r,
                    C = c;
                while (C + 1 < grid[0].length && grid[r][C + 1] === char) {
                    C++;
                }
                while (
                    R + 1 < grid.length &&
                    _.range(c, C + 1).every((idx) => grid[R + 1][idx] === char)
                ) {
                    R++;
                }
                bounds[char] = { r, c, R, C };
            }
        }
    }

    return _.mapValues(bounds, ({ r, c, R, C }) => ({
        row: r,
        col: c,
        height: R - r + 1,
        width: C - c + 1,
    }));
}

type GridLayoutConfig = {
    rowHeight?: 'fit' | 'equal';
    colWidth?: 'fit' | 'equal';
    showLabels?: boolean;
};

const kNoneSpecified = Symbol();

class GridLayoutFragmentAssembler extends FragmentAssembler {
    private _cells: Record<string, GridCell> = {};
    private _items: Record<string, any> = {};
    private _rowHeight?: 'fit' | 'equal';
    private _colWidth?: 'fit' | 'equal';
    private _showLabels?: boolean;

    constructor(
        cells?: string | (GridCell & { name: string })[],
        items?: Record<string, any>,
        config: GridLayoutConfig = {},
    ) {
        super();
        if (cells) {
            if (typeof cells === 'string') {
                // Cells provided as specification string.
                this._cells = parseGridString(cells);
            } else if (Array.isArray(cells)) {
                // Cells already provided in structured format.
                cells.forEach(({ name, ...cell }) => (this._cells[name] = cell));
            } else {
                throw new Error(`Unknown format received for cells: ${cells}`);
            }
        }

        if (items) this._items = items;
        this.config(config);
    }

    public cell(
        name: string,
        row: number,
        col: number,
        height: number,
        width: number,
        item: any = kNoneSpecified,
    ) {
        this._cells[name] = { row, col, height, width };
        if (item !== kNoneSpecified) this._items[name] = item;
        return this;
    }

    public item(name: string, item: any) {
        this._items[name] = item;
        return this;
    }

    public config(config: GridLayoutConfig) {
        const { rowHeight, colWidth, showLabels } = config;
        if (rowHeight !== undefined) this._rowHeight = rowHeight;
        if (colWidth !== undefined) this._colWidth = colWidth;
        if (showLabels !== undefined) this._showLabels = showLabels;
        return this;
    }

    public assemble(getId: (obj: any, name: string) => FragmentId): [GridLayoutFragment, any[]] {
        _.keys(this._cells).forEach((name) => {
            if (!(name in this._items)) throw new Error(`Cell with no item: ${name}`);
        });
        return [
            {
                type: 'GridLayout',
                contents: {
                    cells: _.entries(this._cells).map(([name, cell]) => ({
                        fragmentId: getId(this._items[name], name),
                        ...cell,
                    })),
                    rowHeight: this._rowHeight,
                    colWidth: this._colWidth,
                    showLabels: this._showLabels,
                },
                meta: this._meta,
            },
            Object.keys(this._cells).map((name) => this._items[name]),
        ];
    }
}

/**
 * `Grid` is a layout with slots (named cells) arranged on a 2D grid. Cells can have different
 * dimensions and can even overlap.
 * @param cells
 *     (optional) (1) A list of named cells. (2) A specification string where cells must be
 *     contiguous rectangles of the same character, and all instances of a character must be part
 *     of the same rectangle. Each row is on a different line and all other whitespace is ignored.
 *     The special character '.' is ignored.
 * @param items
 *     (optional) Items to fill slots with.
 * @example
 *     // These produce the equivalent cells.
 *     Grid([{ name: "A", row: 0, col: 0, width: 1, height: 1 },
 *           { name: "B", row: 0, col: 1, width: 3, height: 1 }])
 *     Grid(`A BBB`)
 *     Grid(`A B B B`)
 *
 *     // Use multi-line strings to create rows.
 *     Grid(`A BB C
 *           D BB E`)
 */
export function Grid(
    cells?: string | (GridCell & { name: string })[],
    items?: Record<string, any>,
    config: GridLayoutConfig = {},
) {
    return new GridLayoutFragmentAssembler(cells, items, config);
}

export interface Grid extends ReturnType<typeof Grid> {}

// TODO: Allow width of grid to be set independently (e.g. so that ". A B ." case works).

File: bindings/python/vizstack/assemblers/grid.py
from vizstack.fragment_assembler import FragmentAssembler
from typing import Optional, Tuple, Dict, List, Any, Union
from typing_extensions import TypedDict, Literal
from vizstack.schema import JsonType, View, Fragment
import re


GridCell = TypedDict('GridCell', {
    'col': int,
    'row': int,
    'width': int,
    'height': int,
})

GridCellNamed = TypedDict('GridCellNamed', {
    'col': int,
    'row': int,
    'width': int,
    'height': int,
    'name': str,
})

GridBounds = TypedDict('GridBounds', {
    'r': int,
    'c': int,
    'R': int,
    'C': int,
})

RowColSetting = Literal['fit', 'equal', None]


def _within_bounds(bounds: GridBounds, row: int, col: int) -> bool:
    return bounds['r'] <= row <= bounds['R'] and bounds['c'] <= col <= bounds['C']


def _parse_grid_string(spec: str) -> Dict[str, GridCell]:
    rows = re.split("\||\n", spec)
    rows = ["".join(row.split()) for row in rows]
    rows = [row for row in rows if len(row) > 0]
    if len(rows) == 0 or not all(len(row) == len(rows[0]) for row in rows):
        raise ValueError('Specification string must be rectangular, got rows: ' + str(rows))

    bounds: Dict[str, GridBounds] = {}
    for r in range(len(rows)):
        for c in range(len(rows[0])):
            char = rows[r][c]
            if char in bounds:
                # A previously encountered character must be within the greedily expanded bounds.
                if _within_bounds(bounds[char], r, c): continue
                raise ValueError('Specification string malformed for cell: ' + char)
            else:
                if char == '.': continue
                # Create a new cell and greedily expand col-wise then row-wise while bounds keep
                # fencing a contiguous rectangle of `char`.
                R, C = r, c
                while C + 1 < len(rows[0]) and rows[r][C + 1] == char:
                    C += 1
                while R + 1 < len(rows) and all(rows[R + 1][i] == char for i in range(c, C + 1)):
                    R += 1
                bounds[char] = {'r': r, 'c': c, 'R': R, 'C': C}

    return {name: {
        'row': bound['r'],
        'col': bound['c'],
        'height': bound['R'] - bound['r'] + 1,
        'width': bound['C'] - bound['c'] + 1,
    } for name, bound in bounds.items()}


class Grid(FragmentAssembler):

    _NONE_SPECIFIED = object()

    _row_height: RowColSetting = None
    _col_width: RowColSetting = None
    _show_labels: Optional[bool] = None

    def __init__(self,
                 cells: Optional[Union[str, List[GridCellNamed]]] = None,
                 items: Optional[Dict[str, Any]] = None,
                 row_height: RowColSetting = None,
                 col_width: RowColSetting = None,
                 show_labels: Optional[bool] = None) -> \
            None:
        """
        Args:
            cells: A string like "ABB\nACC\nACC", which specifies initial cell sizes and positions.
                The given example creates a cell "A" at (0,0) with dimensions 1x3, cell "B" at
                (1,0) with dimensions 2x1, and a cell "C" at position (1,1) with dimensions 2x2.
            items: An optional mapping of cell names to items.
        """""
        super(Grid, self).__init__()
        self._cells: Dict[str, GridCell] = dict()
        if isinstance(cells, list):
            for cell in cells:
                self._cells[cell['name']] = {
                    'col': cell['col'],
                    'row': cell['row'],
                    'width': cell['width'],
                    'height': cell['height'],
                }
        elif isinstance(cells, str):
            self._cells = _parse_grid_string(cells)
        else:
            raise ValueError('Unknown format received for cells:' + str(cells))

        self._items: Dict[str, Any] = dict()
        if items is not None:
            for name, item in items.items():
                self.item(name, item)

        self.config(row_height=row_height, col_width=col_width, show_labels=show_labels)

    def cell(self, name: str, row: int, col: int, height: int, width: int, item=_NONE_SPECIFIED):
        self._cells[name] = {
            'col': col,
            'row': row,
            'width': width,
            'height': height,
        }
        if item is not Grid._NONE_SPECIFIED:
            self.item(item, name)
        return self

    def item(self, name: str, item: Any):
        self._items[name] = item
        return self

    def config(self, row_height: RowColSetting = None,
                 col_width: RowColSetting = None,
                 show_labels: Optional[bool] = None):
        if row_height is not None: self._row_height = row_height
        if col_width is not None: self._col_width = col_width
        if show_labels is not None: self._show_labels = show_labels

    def assemble(self, get_id) -> Tuple[Fragment, List[Any]]:
        for cell_name in self._cells:
            assert cell_name in self._items, 'No item was provided for cell "{}".'.format(cell_name)
        return {
                   'type': 'GridLayout',
                   'contents': {
                       'cells': [{**cell, 'fragmentId': get_id(self._items[cell_name], cell_name)}
                                 for cell_name, cell in self._cells.items()],
                       'rowHeight': self._row_height,
                       'colWidth': self._col_width,
                       'showLabels': self._show_labels,
                   },
                   'meta': self._meta,
               }, [self._items[cell_name] for cell_name in self._cells]
